### ISSUE TITLE
Custom predicates in zio.prelude.Assertion

### ISSUE THREAD

**Issue by mijicd** [zio/zio-prelude#783]
Could it be possible to add another constructor, similar to [Assertion.assertion](https://github.com/zio/zio/blob/1bb0368e74f51e7c8b95c8675f731736d15cc527/test/shared/src/main/scala/zio/test/Assertion.scala#L115), that would allow passing custom predicates? If not, what would be the recommended alternative when dealing with "3rd party" validators (e.g., validating URLs)?

cc @kitlangton @adamgfraser

**Comment by kitlangton** [comment]
The only issue with this is that these fully custom assertions may not work at compile-time. By limiting to a known subset, we can reconstitute the methods at compile-time. But I could see this being valuable, even if it couldn't work at compile-time, just for the CustomType.make feature.

**Comment by kitlangton** [comment]
Though perhaps I can mind-meld with @deusaquilus and see if I can copy some more quill-magic to get this to work.

**Comment by mschuwalow** [comment]
@kitlangton 
In scala 2 something like this should be possible by evaling the code, similar to what refined is doing: https://github.com/umbreak/refined/blob/fc0fa21e917986e25a6a164192baa286c05bbe8d/modules/docs/macro_pitfalls.md

This seems to work for Liftable and is enough to have Assertion.Predicate (but produces ugly errors when used incorrectly):
```scala
case q"${A(_)}.predicate[$_](${tree: Tree})" =>
  Assertion.Predicate(c.eval(c.Expr[Function1[A, Boolean]](c.untypecheck(tree.duplicate))), show(tree))

```

For dotty nothing like eval exists and I don't think macros are the way to go there.
Instead, we can rely on inline to simplify the whole assertion to a constant. NewType becomes:
```scala
abstract class NewtypeSmart[A] {
  inline def assertion(inline value: A): AssertionResult
  inline def apply(inline value: A): Type =
    inline assertion(value) match {
      case AssertionResult.Success => value.asInstanceOf[Type]
      case AssertionResult.Failure(err) => fail(err.render(value))
    }
}
```
and assertions are just functions that get fully inlined:
```scala
object Assertion {
  import AssertionResult._

  inline def anything: AssertionResult = Success

  inline def greaterThanOrEqualTo(inline n: Int)(inline value: Int): AssertionResult =
    inline if (value >= n) Success else Failure(AssertionError.Failure("foo"))
}
```
I only got simple cases working in the second approach so far, but in principle it should work for everything we are doing and allows for user defined functions.

If we don't mind breaking backwards compat for dotty I can open a pr with this.

**Comment by kitlangton** [comment]
For Scala 3 we'll certainly be out of luck. But that's great! Perhaps a Scala 2 only extension then?

**Comment by kitlangton** [comment]
We can also add a whole bunch of additional assertions. Let me know if there are any you'd like :)

**Comment by mijicd** [comment]
Scala 2 extension would certainly fix the problem for us at the moment. I'm not sure how we can implement it for Scala 3, though. While more assertions are helpful, I think they transfer the validator maintenance cost to "userland", which is unfortunate and, in certain situations, dangerous.

**Comment by mschuwalow** [comment]
Played around a little bit with scala3. If we are fine with making the user write a bit more code this seems to work:
https://github.com/zio/zio-prelude/pull/793

**Comment by mijicd** [comment]
Closed by @mschuwalow in #793
